microceph/cmd/microceph/disk_add.go

```diff
@@ -64,28 +88,115 @@ func (c *cmdDiskAdd) Run(cmd *cobra.Command, args []string) error {
+		if err != nil {
+			return err
+		}
+
+		// Prepare Batch arguments
+		for _, disk := range disks {
+			req.Path = append(req.Path, disk.Path)
+		}
+	} else {
+		// Pass space separated params as disk paths.
+		req.Path = args
+
+		if !strings.Contains(req.Path[0], common.LoopSpecId) {
+			if c.walDevice != "" {
+				req.WALDev = &c.walDevice
+				req.WALWipe = c.walWipe
+				req.WALEncrypt = c.walEncrypt
+			}
+
+			if c.dbDevice != "" {
+				req.DBDev = &c.dbDevice
+				req.DBWipe = c.dbWipe
+				req.DBEncrypt = c.dbEncrypt
+			}
+		}
 	}
 
-	if c.walDevice != "" {
-		req.WALDev = &c.walDevice
-		req.WALWipe = c.walWipe
-		req.WALEncrypt = c.walEncrypt
+	// required request params.
+	req.Wipe = c.flagWipe
+	req.Encrypt = c.flagEncrypt
+	failures, err := client.AddDisk(context.Background(), cli, &req)
+	if err != nil {
+		return err
 	}
 
-	if c.dbDevice != "" {
-		req.DBDev = &c.dbDevice
-		req.DBWipe = c.dbWipe
-		req.DBEncrypt = c.dbEncrypt
+	err = printAddDiskFailures(failures)
+	if err != nil {
+		return fmt.Errorf("failed to render disk addition output %w", err)
+	}
+
+	return nil
+}
+
+func printAddDiskFailures(responses types.DiskAddResponses) error {
+	var errStr string
+	data := [][]string{}
+
+	if len(responses) == 0 {
+		// No responses; nothing to do.
+		return nil
 	}
 
-	err = client.AddDisk(context.Background(), cli, req)
+	failureCount := 0
+	for _, response := range responses {
+		if strings.Contains(response.Response, "Failure") {
+			failureCount += 1
+			errStr = response.Error
+		}
+		// prepare data for tabulated display.
+		data = append(data, []string{response.Path, response.Response})
+	}
+
+	// Print disk add failures in tabulated form.
+	fmt.Println("")
+	header := []string{"Path", "Status"}
+	sort.Sort(lxdCmd.SortColumnsNaturally(data))
+	err := lxdCmd.RenderTable(lxdCmd.TableFormatTable, header, data, data)
 	if err != nil {
 		return err
 	}
 
+	if failureCount == 1 {
+		// Print error if only one instance of error is there.
+		fmt.Println("Error: ", errStr)
+	} else if failureCount > 1 {
+		// Print error if only one instance of error is there.
+		fmt.Println("Failed adding multiple disks, please check logs for details.")
+	}
+
+	return nil
+}
+
+// validateBatchArgs checks if no loop spec is provided as an argument to batch disk addition.
+func (c *cmdDiskAdd) validateBatchArgs(args []string) error {
+	// no validation if single arg is provided.
+	if len(args) == 1 {
+		return nil
+	}
+
+	// if wal/db devices are provided with batch commands.
+	if c.walDevice != "" {
+		return fmt.Errorf("--wal-device flag is not supported for batch disk addition")
+	}
+
+	if c.dbDevice != "" {
+		return fmt.Errorf("--db-device flag is not supported for batch disk addition")
+	}
+
+	// generate suffix array to look for loop specs in batch args.
+	buf := &bytes.Buffer{}
+	gob.NewEncoder(buf).Encode(args)
+	data := suffixarray.New(buf.Bytes())
+	indices := data.Lookup([]byte(common.LoopSpecId), 1)
```

Comment:
I'm a bit unsure of the use of gob and suffixarray here. Aiui suffixarray would search for any occurence not just prefixes I believe? And would gob encode work with multibyte characters?

I'd suggest to go with a more straightforward approach here, something like

```
for _, a := range args {
    if strings.HasPrefix(a, common.LoopSpecId) {
        return fmt.Errorf("loop spec is not supported as an argument to batch disk addition, use separately")
      }
}
```

Reply:
suffixarrays basically allow searching for strings in list of strings in logarithmic time. But I feel this optimisation isn't really needed yet.